crates/data_components/src/s3_vectors/partition.rs

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use datafusion::{prelude::Expr, scalar::ScalarValue};
+use snafu::prelude::*;
+use twox_hash::XxHash64;
+
+const HASH_SEED: u64 = 7;
+
+const INDEX_NAME_MAX_LENGTH: usize = 25;
+const COLUMN_NAME_MAX_LENGTH: usize = 25;
+const PARTITION_VALUE_MAX_LENGTH: usize = 5;
+const PARTITION_BY_MAX_LENGTH: usize = 5;
+
+static PARTS_SEPARATOR: &str = "-";
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Expected 4 parts in index name between hypens but got {num_parts}"))]
+    IncorrectNumPartsInName { num_parts: usize },
+}
+
+#[derive(Debug)]
+pub struct PartitionedIndexName {
+    index_name: String,
+    column_name: String,
+    partition_value_hash: String,
+    partition_by_hash: String,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum BelongsWith {
+    SameDataset,
+    DifferentDataset,
+    DifferentParitionByExpressions,
+}
+
+impl PartitionedIndexName {
+    pub fn new(
+        index_name: &str,
+        column_name: &str,
+        partition_value: &ScalarValue,
+        partition_by: &[Expr],
+    ) -> Result<Self, Error> {
+        let index_name = truncate(&sanitize(index_name), INDEX_NAME_MAX_LENGTH);
+        let column_name = truncate(&sanitize(column_name), COLUMN_NAME_MAX_LENGTH);
+        let partition_value_hash = truncate(
+            &hash_to_hex(&partition_value.to_string()),
+            PARTITION_VALUE_MAX_LENGTH,
+        );
+        let partition_by_hash = truncate(
+            &hash_to_hex(&format!("{partition_by:?}")),
+            PARTITION_BY_MAX_LENGTH,
+        );
+        Ok(Self {
+            index_name,
+            column_name,
+            partition_value_hash,
+            partition_by_hash,
+        })
+    }
+
+    /// Format an index name suitable for S3 Vectors
+    #[must_use]
+    pub fn to_index_name(&self) -> String {
+        [
+            self.index_name.clone(),
+            self.column_name.clone(),
+            self.partition_value_hash.clone(),
+            self.partition_by_hash.clone(),
+        ]
+        .join(PARTS_SEPARATOR)
+    }
+
+    pub fn from_index_name(index_name: &str) -> Result<Self, Error> {
+        let parts: Vec<&str> = index_name.split('-').collect();
```

Comment:
I think this will need to handle the case where the index_name/column_name have dashes

Reply:
Either that, or sanitize the index_name/column_name to use something other than a dash if they have one.

Reply:
you really woudn't be able to tell which part (index or column) contains the hyphen in the case where there is an extra hyphen. I think a different separator would be better. I think a `.` would work: https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-vectors-indexes.html#s3-vectors-indexes-naming and it is valid for filenames in filesystems.